Lib/test/test_complex.py

```diff
@@ -351,6 +354,45 @@ def test_pow_with_small_integer_exponents(self):
                     self.assertEqual(str(float_pow), str(int_pow))
                     self.assertEqual(str(complex_pow), str(int_pow))
 
+        # Check that complex numbers with special components
+        # are correctly handled.
+        values = [complex(*_) for _ in combinations([1, -1, 0.0, -0.0, 2,
+                                                     -3, INF, -INF, NAN], 2)]
+        exponents = [0, 1, 2, 3, 4, 5, 6, 19]
+        for z in values:
+            for e in exponents:
+                try:
+                    r_pow = z**e
+                except OverflowError:
+                    continue
+                r_pro = reduce(lambda x, y: x*y, [z]*e) if e else 1+0j
+                if str(r_pow) == str(r_pro):
+                    continue
+
+                self.assertNotIn(z.real, {0.0, -0.0, INF, -INF, NAN})
+                self.assertNotIn(z.imag, {0.0, -0.0, INF, -INF, NAN})
+
+                # We might fail here, because associativity of multiplication
+                # is broken already for floats.
```

Comment:
```suggestion
                # is already broken for signed zeroes.
```

I think only the sign of zero may have a different behaviour. Otherwise, associativity shouldn't be broken, right?

Reply:
No, it's damn easy to break associativity for normal floats.  A quick (random) example:
```pycon
>>> a, b, c = 1.123+1.2j, 3.2+1.41j, 1.1+2j
>>> (a*b)*c
(-8.755099999999999+9.768973j)
>>> a*(b*c)
(-8.755099999999999+9.768973000000003j)
>>> ((a*a)*a)*a
(-7.232108205358999-0.9641862383999995j)
>>> (a*a)*(a*a)
(-7.232108205358999-0.9641862384j)
```

Reply:
Huh... I've been lied to for all those years.... Well, it's within the realm of numerical imprecision (which I am quite used to when working with gradient descents & co...) so it's not *that* bad :') (well still bad but not that catastrophically bad as I imagine when I read "broken associativity").

Reply:
Well, it's exactly *that* bad: ``(a*b)*c != a*(b*c)``, this algebraic identity is broken.

Reply:
Yes, algebraically speaking, but it's up to some minor imprecision. Though, maybe the imprecision will be carried over and amplified? (in general, I would never compare two non-integral values using ==, but rather "isclose").

Reply:
> the imprecision will be carried over and amplified?

Simple example with addition:
```pycon
>>> 1e10 + 1e100 + 1.0 + -1e100
0.0
>>> 1e100 + -1e100 + 1e10 + 1.0  # exact sum
10000000001.0
```

Reply:
Oh yes, that's right! I remember this now. Urgh.